### ISSUE TITLE
Step 3: Initial Feature Extraction - Standard Conv Layer Implementation

### ISSUE THREAD

**Issue by SCSBalaji** [SCSBalaji/birthday_remainder_project#29]
## Objective
Implement the first stage of the Modified MobilePlantViT architecture: Initial Feature Extraction using a Standard 3×3 Convolution layer. This layer takes the input image (3×224×224) and produces feature maps (16×112×112).

---

## Dependencies
- ✅ Issue #18 (Dataset Collection)
- ✅ Issue #19 (Data Preprocessing Pipeline)

---

## Architecture Reference
According to your architecture diagram, **Stage 2** performs:
- **Input:** 3×224×224 (RGB image)
- **Operation:** Standard Conv 3×3, Stride=2, Padding=1
- **Output:** 16×112×112 (feature maps)
- **Purpose:** Initial spatial reduction and channel expansion

---

### Step 1: Create Project Structure

```bash
# Create folders for model components
mkdir -p src/models/layers
mkdir -p src/models/blocks
mkdir -p tests
mkdir -p config

# Create __init__.py files
touch src/models/__init__.py
touch src/models/layers/__init__.py
touch src/models/blocks/__init__.py
```

---

### Step 2: Create Model Configuration

Create `config/model_config.yaml`:
```yaml
# Model Configuration for Modified MobilePlantViT

model:
  name: "Modified-MobilePlantViT"
  version: "1.0"
  
  input:
    size: 224
    channels: 3
  
  initial_conv:
    in_channels: 3
    out_channels: 16
    kernel_size: 3
    stride: 2
    padding: 1
    bias: false
    activation: "relu"
```

---

### Step 3: Implement Standard Convolution Layer

Create `src/models/layers/conv_layers.py`:
```python
"""
Standard Convolution Layers for Modified MobilePlantViT
"""

import torch
import torch.nn as nn


class StandardConv2d(nn.Module):
    """
    Standard 2D Convolution with BatchNorm and Activation
    
    Args:
        in_channels: Number of input channels
        out_channels: Number of output channels
        kernel_size: Size of convolution kernel
        stride: Stride for convolution
        padding: Padding for convolution
        activation: Activation function ('relu', 'relu6', 'silu', or None)
    """
    
    def __init__(self, in_channels, out_channels, kernel_size=3, 
                 stride=1, padding=0, activation='relu'):
        super(StandardConv2d, self).__init__()
        
        # Convolution layer
        self.conv = nn.Conv2d(in_channels, out_channels, kernel_size, 
                              stride, padding, bias=False)
        
        # Batch Normalization
        self.bn = nn.BatchNorm2d(out_channels)
        
        # Activation function
        if activation == 'relu':
            self.activation = nn.ReLU(inplace=True)
        elif activation == 'relu6':
            self.activation = nn.ReLU6(inplace=True)
        elif activation == 'silu':
            self.activation = nn.SiLU(inplace=True)
        else:
            self.activation = nn.Identity()
    
    def forward(self, x):
        x = self.conv(x)
        x = self.bn(x)
        x = self.activation(x)
        return x


class InitialFeatureExtraction(nn.Module):
    """
    Initial Feature Extraction Stage
    Input: (B, 3, 224, 224) → Output: (B, 16, 112, 112)
    """
    
    def __init__(self, in_channels=3, out_channels=16):
        super(InitialFeatureExtraction, self).__init__()
        
        self.conv = StandardConv2d(
            in_channels=in_channels,
            out_channels=out_channels,
            kernel_size=3,
            stride=2,  # Spatial reduction: 224 → 112
            padding=1,
            activation='relu'
        )
    
    def forward(self, x):
        return self.conv(x)


# Quick test
if __name__ == "__main__":
    print("Testing Initial Feature Extraction...")
    model = InitialFeatureExtraction()
    x = torch.randn(1, 3, 224, 224)
    out = model(x)
    print(f"Input shape: {x.shape}")
    print(f"Output shape: {out.shape}")
    print(f"Expected: torch.Size([1, 16, 112, 112])")
    assert out.shape == torch.Size([1, 16, 112, 112])
    print("✅ Test passed!")
```

---

### Step 4: Create Test Suite

Create `tests/test_conv_layer.py`:
```python
"""
Unit tests for Standard Convolution Layer
"""

import torch
import sys
from pathlib import Path

sys.path.append(str(Path(__file__).parent.parent))

from src.models.layers.conv_layers import StandardConv2d, InitialFeatureExtraction


def test_initial_feature_extraction():
    """Test InitialFeatureExtraction module"""
    print("="*70)
    print("Testing Initial Feature Extraction")
    print("="*70)
    
    model = InitialFeatureExtraction()
    
    # Test 1: Standard input
    print("\n[Test 1] Standard input (224×224)")
    x = torch.randn(4, 3, 224, 224)
    out = model(x)
    assert out.shape == (4, 16, 112, 112)
    print(f"✅ Input: {x.shape} → Output: {out.shape}")
    
    # Test 2: Different batch sizes
    print("\n[Test 2] Batch size variations")
    for batch_size in [1, 8, 32]:
        x = torch.randn(batch_size, 3, 224, 224)
        out = model(x)
        assert out.shape == (batch_size, 16, 112, 112)
        print(f"✅ Batch size {batch_size}: {out.shape}")
    
    # Test 3: Parameter count
    print("\n[Test 3] Parameter count")
    total = sum(p.numel() for p in model.parameters())
    trainable = sum(p.numel() for p in model.parameters() if p.requires_grad)
    print(f"Total parameters: {total:,}")
    print(f"Trainable parameters: {trainable:,}")
    
    # Test 4: Gradient flow
    print("\n[Test 4] Gradient flow")
    x = torch.randn(2, 3, 224, 224, requires_grad=True)
    out = model(x)
    loss = out.sum()
    loss.backward()
    assert x.grad is not None
    print("✅ Gradients flowing correctly")
    
    # Test 5: Device compatibility
    print("\n[Test 5] Device compatibility")
    if torch.cuda.is_available():
        model_gpu = model.cuda()
        x_gpu = torch.randn(1, 3, 224, 224).cuda()
        out_gpu = model_gpu(x_gpu)
        print(f"✅ GPU mode: {out_gpu.device}")
    else:
        print("⚠️  GPU not available")
    
    print("\n" + "="*70)
    print("🎉 ALL TESTS PASSED!")
    print("="*70)


if __name__ == "__main__":
    test_initial_feature_extraction()
```

---

## Acceptance Criteria
- [ ] Project structure created
- [ ] `config/model_config.yaml` created
- [ ] `StandardConv2d` layer implemented
- [ ] `InitialFeatureExtraction` module implemented
- [ ] All unit tests pass
- [ ] Model outputs correct shape: (B, 16, 112, 112)
- [ ] Parameter count is reasonable (~464 parameters)
- [ ] Gradient flow works correctly

---

## Testing Commands
```bash
# Test the layer directly
python src/models/layers/conv_layers.py

# Run unit tests
python tests/test_conv_layer.py
```

---

## Estimated Time: 2-3 hours

---

## Next Step
Once completed, Issue #4 will implement GhostConv Blocks (Stage 3).
